cairo_zero/kakarot/precompiles/precompiles_helpers.cairo

```diff
@@ -87,9 +87,18 @@ namespace PrecompilesHelpers {
         alloc_locals;
         let precompile_requires_whitelist = requires_whitelist(precompile_address);
 
+        tempvar is_not_delegatecall: felt = Helpers.is_zero(message_address - precompile_address);
+
         // Ensure that calls to precompiles that require a whitelist are properly authorized.
         if (precompile_requires_whitelist == TRUE) {
-            let is_whitelisted = is_caller_whitelisted(caller_address);
+            // If the call is not a DELEGATECALL / CALLCODE, the actual caller is the caller_address
+            // Otherwise, the actual caller is the message_address.
+            if (is_not_delegatecall != FALSE) {
+                tempvar actual_caller_address = caller_address;
+            } else {
+                tempvar actual_caller_address = message_address;
+            }
+            let is_whitelisted = is_caller_whitelisted(actual_caller_address);
```

Comment:
can you explain the reasoning behind this? im not sure I get it

Reply:
The idea of the cairo fix is to whitelist on the code that is calling the precompile. In this case the dualVmToken.
In case of a normal call: the address of the dualVMtoken is the caller_address (the address that called the address of execution)
`tempvar caller_address = evm.message.caller;`

In case of delegatecall, the latter is changed. The `evm.message.caller` is now the one that called the dualVmTokens that calls the precompiles.
And the address of the dualVMtokens is `evm.message.address.evm`

From `exec_precompile`
```
@param caller_address The address of the caller of the precompile. Delegatecall rules apply.
@param message_address The address being executed in the current message.
```
In fact delegate rules applies also for the `message_address`